src/WignerSymbols.jl

```diff
@@ -242,6 +247,85 @@ function racahW(T::Type{<:Real}, j₁, j₂, J, j₃, J₁₂, J₂₃)
     end
 end
 
+"""
+    wigner9j(T::Type{<:Real} = RationalRoot{BigInt}, j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉) -> ::T
+
+Compute the value of the Wigner-9j symbol
+⎧ j₁ j₂ j₃ ⎫
+⎨	j₄ j₅ j₆ ⎬
+⎩	j₇ j₈ j₉ ⎭
+as a type `T` real number. By default `T = RationalRoot{BigInt}`.
+
+Returns `zero(T)` if any of the triangle conditions TODO 
+are not satisfied, but throws a `DomainError` if
+the `jᵢ`s are not integer or halfinteger.
+"""
+wigner9j(j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉) = wigner9j(RRBig, j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉)
+function wigner9j(T::Type{<:Real}, j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉) 
+    # check validity of `jᵢ`s
+    for jᵢ in (j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉)
+        (ishalfinteger(jᵢ) && jᵢ >= zero(jᵢ)) || throw(DomainError("invalid jᵢ", jᵢ))
+    end
+    return _wigner9j(T, HalfInteger.((j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉))...)
+end
+
+const _perms9j = [([1, 2, 3], [1, 2, 3]) ([1, 2, 3], [1, 3, 2]) ([1, 2, 3], [2, 1, 3]) ([1, 2, 3], [2, 3, 1]) ([1, 2, 3], [3, 1, 2]) ([1, 2, 3], [3, 2, 1]);
+                  ([1, 3, 2], [1, 2, 3]) ([1, 3, 2], [1, 3, 2]) ([1, 3, 2], [2, 1, 3]) ([1, 3, 2], [2, 3, 1]) ([1, 3, 2], [3, 1, 2]) ([1, 3, 2], [3, 2, 1]);
+                  ([2, 1, 3], [1, 2, 3]) ([2, 1, 3], [1, 3, 2]) ([2, 1, 3], [2, 1, 3]) ([2, 1, 3], [2, 3, 1]) ([2, 1, 3], [3, 1, 2]) ([2, 1, 3], [3, 2, 1]);
+                  ([2, 3, 1], [1, 2, 3]) ([2, 3, 1], [1, 3, 2]) ([2, 3, 1], [2, 1, 3]) ([2, 3, 1], [2, 3, 1]) ([2, 3, 1], [3, 1, 2]) ([2, 3, 1], [3, 2, 1]);
+                  ([3, 1, 2], [1, 2, 3]) ([3, 1, 2], [1, 3, 2]) ([3, 1, 2], [2, 1, 3]) ([3, 1, 2], [2, 3, 1]) ([3, 1, 2], [3, 1, 2]) ([3, 1, 2], [3, 2, 1]);
+                  ([3, 2, 1], [1, 2, 3]) ([3, 2, 1], [1, 3, 2]) ([3, 2, 1], [2, 1, 3]) ([3, 2, 1], [2, 3, 1]) ([3, 2, 1], [3, 1, 2]) ([3, 2, 1], [3, 2, 1])]
+
+const _signs9j = [1 -1 -1 1 1 -1;
+                  -1 1 1 -1 -1 1;
+                  -1 1 1 -1 -1 1;
+                  1 -1 -1 1 1 -1;
+                  1 -1 -1 1 1 -1;
+                  -1 1 1 -1 -1 1]
+
+function _wigner9j(T::Type{<:Real}, j₁::HalfInteger, j₂::HalfInteger, j₃::HalfInteger, 
+                                    j₄::HalfInteger, j₅::HalfInteger, j₆::HalfInteger,
+                                    j₇::HalfInteger, j₈::HalfInteger, j₉::HalfInteger)
+  
+    # check triangle conditions 
+    if !(δ(j₁, j₂, j₃) && δ(j₄, j₅, j₆) && δ(j₇, j₈, j₉) && δ(j₁, j₄, j₇) && δ(j₂, j₅, j₈) && δ(j₃, j₆, j₉))
+        return zero(T)
+    end
+   
+    # dictionary lookup, check all 72 permutations 
+    k = [j₁ j₂ j₃; j₄ j₅ j₆; j₇ j₈ j₉]
+    for (p, m) in zip(_perms9j, _signs9j)
+        kk = Tuple(reshape(k[p...], 9))
```

Comment:
I have to admit that I am confused by how the permutations are encoded here and so how this code functions exactly. Would it be possible to follow the 3j and 6j philosophy and try to impose a canonical order. For example, given all the symmetries of the 9j symbol, it seems like one could always impose:

`j₁ <= j₂ <= j₄ <= min(j₃, j₅, j₆, j₇, j₈, j₉)`

When the inequalities are strict, I believe this fixes the order completely, i.e. any permuation or reflection would destroy this order. However, when there are equalities, further conditions could be imposed.

Reply:
I was following the approach mentioned by the [wigxjpf](https://fy.chalmers.se/subatom/wigxjpf/) authors in [this slide-deck](https://fy.chalmers.se/~f96hajo/shows/htj_nov2015_wigxj.pdf) (3rd slide from the end), where the symmetry checks for the 9j are done explicitly for the 72 equivalent permutations+transposition (up to sign) against the cache, rather than using [a canonical ordering like the 3j or 6j](https://epubs.siam.org/doi/10.1137/S1064827503422932).
 
The rationale there being (I believe) that unlike what happens in the 3j or 9j cases, the magic square symmetries give 6 conditions on 9 linearly independent entries rather than over-constraining the system to give a single canonical ordering.

My implementation could probably be cleaner or made more readable, but I'm using that Julia arrays can be indexed with vectors to get a permutation, e.g. 
```julia
A = ['a', 'b', 'c']

A[[2, 3, 1]] # = ['b', 'c', 'a']

B = [1 2 3; 4 5 6; 7 8 9]

# swap the first two rows, and last two columns
B[[2, 1, 3], [1, 3, 2]] # = [4 6 5; 1 3 2; 7 9 8]
```
and then hard-coded all the column and row permutations accordingly (in lexicographic order) as well as the parity of each permutation. The transpositions get checked too but aren't encoded in the `_perms9j` table of indices, hence the `kk` and `kkT` cases.

I think you're right though; breaking it into a few cases should be able to do better than checking all permutations. I'm out of time to spare today, so I'll sit down with this one later in the week to work out the cases and see exactly how it can be improved!

Reply:
Great. Looking at the slides, I indeed remember that for 3j and 6j you can go further and really compute a unique linear index for each of the symbols (such that those that are related up to a sign by symmetries have indeed the same linear index). That's probably too much to ask for the 9j symbols, but I don't use that anyway. I simply use the symmetries to reduce them to canonical order, but then still rely on a hash table.

Reply:
I've been busier than planned, so I've only been able to take a brief of a stab at a canonical form. The condition you suggest as a starting point unfortunately doesn't work: *if* a symbol can be canonicalized to meet it, then it's a good condition in that any row or col swap or transposition would break it - however, not all symbols can be canonicalized to satisfy that condition in the first place.

The issue is that anything that sets a particular entry over-constrains the permutations that can follow. e.g. in the suggested canonicalization `j₁` must be made the smallest number in the symbol but doing so disallows the 1st row and 1st column from being involved in future row or column swaps to set the other entries, but in order to move the 2nd smallest number into `j₂` it must have landed in the 1st row or the 1st column during the permutation that set `j₁` - which there's no guarantee it will have. Because of patterns like this I'm finding most canonicalization schemes I can come up with in the same vein require specifying many conditions and gets a bit ugly (and aren't looking much better than just checking the permutations).

I'll see if I can find some better way to approach canonicalization over the weekend here - apologies for the delay.